### PR TITLE
PMM-7346: only positive duration

### PR DESCRIPTION
**Pull request by fabio-silva**
What this PR does / why we need it: Disallows negative durations for Alert Rules

Which issue(s) this PR fixes: https://jira.percona.com/browse/PMM-7346

FB https://github.com/Percona-Lab/pmm-submodules/pull/1434

### CODE REVIEW

**Comment by atymchuk** [comment]
@fabio-silva Good job! Can you provide a link to a feature build?

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx [hunk 1]
```diff
@@ -18,31 +17,97 @@ jest.mock('app/core/app_events', () => {
...
-    await act(async () => {
-      wrapper = mount(<AddAlertRuleModal setVisible={jest.fn()} isVisible={false} />);
-    });
+    expect(wrapper.find(dataQa('add-alert-rule-modal')).exists()).toBeTruthy();
+    expect(wrapper.find(dataQa('add-alert-rule-modal-form')).exists()).toBeTruthy();
+    expect(wrapper.find(dataQa('add-alert-rule-modal-add-button')).exists()).toBeTruthy();
+  });
 
-    expect(wrapper.find(dataQa('add-alert-rule-modal-form')).length).toBe(0);
+  it('doesn not render the modal when visible is set to false', async () => {
```
Comment (tiagomotasantos) [inline public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx:60]:
```suggestion
  it('does not render the modal when visible is set to false', async () => {
```

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx [hunk 2]
```diff
@@ -18,31 +17,97 @@ jest.mock('app/core/app_events', () => {
...
+    const wrapper = mount(<AddAlertRuleModal setVisible={jest.fn()} isVisible />);
+    const button = wrapper.find(dataQa('add-alert-rule-modal-add-button')).find('button');
+
+    expect(button.getDOMNode().hasAttribute('disabled')).toBeTruthy();
+  });
+
+  it('should enable the submit button if all fields are valid', () => {
+    const wrapper = mount(<AddAlertRuleModal setVisible={jest.fn()} isVisible alertRule={initialValues} />);
+
+    const thresholdInput = wrapper.find('#input-threshold-id');
```
Comment (tiagomotasantos) [inline public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx:82]: To be consistent I would use data-qa here instead of id.

**Review by tiagomotasantos — CHANGES_REQUESTED** [review]
Minor things from my side, very good job!

**Comment by fabio-silva** [comment]
@atymchuk Added the FB IP to the initial comment 👍

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx (hunk 2 above)
Comment (atymchuk) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx:82]: Actually, we are not expected to use ids.

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx (hunk 1 above)
Comment (fabio-silva) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx]: Thanks!

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx (hunk 2 above)
Comment (fabio-silva) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx]: You are right. I thought there was no data-qa in the native input, actually there is, thank you!

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.ts [hunk 3]
```diff
@@ -192,3 +192,6 @@ export const getInitialValues = (alertRule: AlertRule): AddAlertRuleFormValues =
     threshold: formatEditThreshold(params),
   };
 };
+
+export const minValidator = (min: number) => (value: any): undefined | string =>
+  value >= min ? undefined : `Must be greather than or equal to ${min}`;
```
Comment (tiagomotasantos) [inline public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.ts:197]:
```suggestion
  value >= min ? undefined : `Must be greater than or equal to ${min}`;
```

**Review by tiagomotasantos — CHANGES_REQUESTED** [review]

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx [hunk 4]
```diff
@@ -18,31 +17,97 @@ jest.mock('app/core/app_events', () => {
...
+    const button = wrapper.find(dataQa('add-alert-rule-modal-add-button')).find('button');
 
-    expect(wrapper.find(dataQa('add-alert-rule-modal-form')).length).toBe(1);
+    expect(button.getDOMNode().hasAttribute('disabled')).toBeFalsy();
+  });
+
+  it('should disable the submit button if a negative duration is inserted', () => {
+    const wrapper = mount(<AddAlertRuleModal setVisible={jest.fn()} isVisible alertRule={initialValues} />);
+    const thresholdInput = wrapper.find('#input-threshold-id');
+    const durationInput = wrapper.find('#input-duration-id');
```
Comment (nicolalamacchia) [inline public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx]: As @tiagomotasantos already mentioned, please avoid using ids

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRulesModal.constants.ts [hunk 5]
```diff
@@ -20,3 +20,6 @@ export const SEVERITY_OPTIONS = [
     label: Messages.severities.SEVERITY_WARNING,
   },
 ];
+
+// Seconds
```
Comment (nicolalamacchia) [inline public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRulesModal.constants.ts:24]: Do we need to know that those are seconds? :)

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.ts (hunk 3 above)
Comment (nicolalamacchia) [inline public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.ts:197]: JFYI: since you have access to the inner HTML input element, I think you could just pass a `min` prop to it, instead of using a validator

**Review by nicolalamacchia — CHANGES_REQUESTED** [review]

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.ts (hunk 3 above)
Comment (tiagomotasantos) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.ts:197]: I thought the same here but I tried this and in some browsers (e.g. firefox) the min doesn't work that well so I would keep this as safe guard.

Comment (fabio-silva) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.ts]: Boy, what a typo 🤦

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRulesModal.constants.ts (hunk 5 above)
Comment (fabio-silva) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRulesModal.constants.ts]: yeah, maybe units don't really matter here

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.ts (hunk 3 above)
Comment (fabio-silva) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.ts]: @nicolalamacchia Yes, I'm passing that in `AddAlertRulesModal` as well. The problem is that with only that, the button will still be enabled. Once you click it, you'll see a native warning on the input, which would be very different from the error labels we have in place now.

**Review by tiagomotasantos — DISMISSED** [review]

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx [hunk 6]
```diff
@@ -18,31 +17,97 @@ jest.mock('app/core/app_events', () => {
...
+    const thresholdInput = wrapper.find(dataQa('threshold-text-input'));
+    thresholdInput.simulate('change', {
+      target: {
+        value: '2',
+      },
     });
+    const button = wrapper.find(dataQa('add-alert-rule-modal-add-button')).find('button');
 
-    expect(wrapper.find(dataQa('add-alert-rule-modal-form')).length).toBe(1);
+    expect(button.getDOMNode().hasAttribute('disabled')).toBeFalsy();
```
Comment (atymchuk) [inline public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx:90]: `getDOMNode().hasAttribute()` seems a bit verbose, why not use `.props('disabled')`? Would it work?

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.tsx [hunk 7]
```diff
@@ -111,7 +112,12 @@ export const AddAlertRuleModal: FC<AddAlertRuleModalProps> = ({ isVisible, setVi
 
             <TextInputField label={Messages.thresholdField} name="threshold" />
 
-            <NumberInputField label={Messages.durationField} name="duration" validators={[required]} />
+            <NumberInputField
+              inputProps={{ min: MINIMUM_DURATION_VALUE }}
```
Comment (atymchuk) [inline public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.tsx:116]: Wouldn't this produce a native validation error, which is not something we expect to see on UI?

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.tsx [hunk 8]
```diff
@@ -111,7 +112,12 @@ export const AddAlertRuleModal: FC<AddAlertRuleModalProps> = ({ isVisible, setVi
...
+              label={Messages.durationField}
+              name="duration"
+              validators={[required, minValidator(MINIMUM_DURATION_VALUE)]}
```
Comment (atymchuk) [inline public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.tsx:119]: Consider taking the validators array out of the component to avoid re-creation on every render.

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx (hunk 2 above)
Comment (atymchuk) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx]: Well, the point is you cannot guarantee the uniqueness of ids, therefore their use is considered a potential source of trouble (or bugs).

**Review by atymchuk — CHANGES_REQUESTED** [review]
I left some comments, pls see if you can address them :)

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx (hunk 6 above)
Comment (fabio-silva) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx]: nice one 👍

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.tsx (hunk 7 above)
Comment (fabio-silva) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.tsx]: Yeah, it would. I added it as a "safeguard", but I believe it won't do anything since the button will be disabled due to the validator. I'm gonna remove it.

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.tsx (hunk 8 above)
Comment (fabio-silva) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.tsx]: 👍

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.ts (hunk 3 above)
Comment (nicolalamacchia) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.ts]: yes, it's still better to have our own validator

**Review by nicolalamacchia — APPROVED** [review]

**Review by tiagomotasantos — DISMISSED** [review]

**Review by nicolalamacchia — APPROVED** [review]

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.test.ts [hunk 9]
```diff
@@ -373,4 +374,11 @@ describe('AddAlertRuleModal utils', () => {
       ])
     ).toBeNull();
   });
+
+  test('minimum validator', () => {
+    expect(minValidator(1)(-1)).not.toBeUndefined();
+    expect(minValidator(1)(0)).not.toBeUndefined();
+    expect(minValidator(1)(1)).toBeUndefined();
+    expect(minValidator(1)(100)).toBeUndefined();
+  });
```
Comment (nicolalamacchia) [inline public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.test.ts:383]: Nit: consider checking edge cases, such as `Infinity`, negative numbers, also in the closure argument

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx [hunk 10]
```diff
@@ -18,31 +17,97 @@ jest.mock('app/core/app_events', () => {
...
+
+    durationInput.simulate('change', {
+      target: {
+        value: '-10',
+      },
+    });
+    const button = wrapper.find(dataQa('add-alert-rule-modal-add-button')).find('button');
 
-    wrapper.unmount();
+    expect(button.props().disabled).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx]: done 👍

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.test.ts (hunk 9 above)
Comment (fabio-silva) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.utils.test.ts]: Good idea 👍

**Review by nicolalamacchia — DISMISSED** [review]

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx [hunk 11]
```diff
@@ -18,31 +17,97 @@ jest.mock('app/core/app_events', () => {
...
+    },
+  };
+
   afterEach(() => {
     jest.clearAllMocks();
   });
 
-  it('doesn not render the modal when visible is set to false', async () => {
-    let wrapper: ReactWrapper<{}, {}, any>;
+  it('should correctly render modal', () => {
```
Comment (atymchuk) [inline public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx:52]: nit: `Should render the modal`, since we the reader wouldn't know what `correctly` means, nor do we want them to guess. The goal of a test is that the component does it the way we programmed it to, so `correctly` is always meant.

**Review by atymchuk — APPROVED** [review]

File: public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx (hunk 11 above)
Comment (fabio-silva) [reply public/app/features/integrated-alerting/components/AlertRules/AddAlertRuleModal/AddAlertRuleModal.test.tsx]: Definitely ;)

**Review by tiagomotasantos — APPROVED** [review]
